include/onnxruntime/core/framework/sparse_tensor.h

```diff
@@ -1,72 +1,317 @@
+  return static_cast<SparseFormatFlags>(static_cast<uint32_t>(flags) & ~static_cast<uint32_t>(flag));
+}
+
+inline bool IsSet(SparseFormatFlags flags, SparseFormatFlags flag) noexcept {
+  return SparseFormatFlags::kUndefined != (flags & flag);
+}
+
+inline SparseFormatFlags Set(SparseFormatFlags flags, SparseFormatFlags flag) noexcept {
+  return flags | flag;
+}
+
+/**
+ * @brief This is a base class for virtual sparse format representation
+ * 
+ */
+class SparseRep {
+ protected:
+  SparseRep(MLDataType data_type, const TensorShape& values_shape, int64_t total_buffer_size, const AllocatorPtr& allocator);
+  SparseRep(MLDataType data_type, const TensorShape& values_shape, void* values, const OrtMemoryInfo& mem_info);
+
+  void* IndicesStart(int64_t align) {
+    return reinterpret_cast<uint8_t*>(values_.MutableDataRaw()) + Roundup(values_.SizeInBytes(), align);
+  }
+
+ public:
+  virtual ~SparseRep();
+
+  ORT_DISALLOW_COPY_AND_ASSIGNMENT(SparseRep);
+
+  const Tensor& Values() const noexcept {
+    return values_;
+  }
+
+  Tensor& MutableValues() noexcept {
+    return values_;
+  }
+
+  bool OwnBuffer() const noexcept {
+    return allocator_ != nullptr;
+  }
+
+  const OrtMemoryInfo& Location() const {
+    return values_.Location();
+  }
+
+  // Round up size to a multiple of align.
+  // Example:
+  //   Roundup(13, 5)   => 15
+  //   Roundup(201, 16) => 208
+  static int64_t Roundup(int64_t size, int64_t align) {
+    return ((size + align - 1) / align) * align;
+  }
+
+  /// <summary>
+  /// Calculate required buffer size. We will place indices
+  /// after data and make sure indices start at int64_t aligned place
+  /// </summary>
+  /// <returns></returns>
+  static int64_t CalculateRequiredBufferSize(int64_t data_size, int64_t indices_size, int64_t align) {
+    return Roundup(data_size, align) + indices_size;
+  }
+
+  /// <summary>
+  ///  Copy the same format to a different destination.
+  /// </summary>
+  /// <param name="data_transfer_manager">manager</param>
+  /// <param name="allocator">allocator for the destination</param>
+  /// <param name="exec_q_id"></param>
+  /// <param name="dst_rep">[out] destination representation</param>
+  /// <returns>status</returns>
+  virtual Status Copy(const DataTransferManager& data_transfer_manager, const AllocatorPtr& allocator,
+                      int exec_q_id, std::unique_ptr<SparseRep>& dst_rep) const = 0;
+
+  virtual Status Copy(const IDataTransfer& data_transfer, const AllocatorPtr& allocator, int exec_q_id,
+                      std::unique_ptr<SparseRep>& dst_rep) const = 0;
+
+  /// <summary>
+  /// Calculates and returns how much this fully initialized SparseTensor data (would)
+  /// occupy in a contiguous allocation block
+  /// </summary>
+  /// <returns></returns>
+  virtual int64_t RequiredAllocationSize() const noexcept = 0;
+
+ private:
+  void ReleaseBuffer();
+
+  AllocatorPtr allocator_;  // Used to allocate p_data_ or nullptr
+  void* p_data_;
```

Comment:
What is p_data_ and how does it relate to values_ ?

Reply:
p_data_ is a buffer where I put everything, values and indices. A common buffer makes it easier to hook it up to memory allocation planner and potentially re-use it.

Reply:
I would like to move p_data_ and values into SparseTensor class and leave indices in the rep. This way, it would be easier to pre-allocate buffer given the pre-calculated size.
In general, there might be 2 allocation strategies:
1) Total size is not known, in this case, we simply instantiate SparseTensor with an allocator and return it to the kernel,
2) the size is known, then the size is allocated Or a suitable buffer for re-use is set.

Reply:
Moving p_data_ and values_ (renaming to sparse_values_ maybe) into SparseTensor makes sense to me.